### PR TITLE
Sac Algorithm

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
+ Unit tested
+ Also tested with the sac_pendulum.gin with batch_env_temp, now all curves matches:
  1) return curve
  2) actor loss
  3) critic loss
  4) alpha loss
+ Further tested discrete action case with sac_cart_pole.gin, performs similar to TF version

### CODE REVIEW

File: alf/algorithms/sac_algorithm_test.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        eval_time_step = unroll(eval_env, alg2, steps_per_episode - 1)
+        print(eval_time_step.reward.mean())
+
+        self.assertAlmostEqual(
+            1.0, float(eval_time_step.reward.mean()), delta=5e-1)
+
+
+if __name__ == '__main__':
+    SACAlgorithmTestDiscrete().test_sac_algorithm_discrete()
+    #unittest.main()
```
Comment (emailweixu) [inline alf/algorithms/sac_algorithm_test.py]: alf.test.main()

File: alf/algorithms/sac_algorithm.py [hunk 2]
```diff
@@ -119,23 +126,18 @@ def __init__(self,
                 networks.
             target_update_period (int): Period for soft update of the target
                 networks.
-            dqda_clipping (float): when computing the actor loss, clips the
-                gradient dqda element-wise between [-dqda_clipping, dqda_clipping].
-                Does not perform clipping if dqda_clipping == 0.
-            actor_optimizer (tf.optimizers.Optimizer): The optimizer for actor.
-            critic_optimizer (tf.optimizers.Optimizer): The optimizer for critic.
-            alpha_optimizer (tf.optimizers.Optimizer): The optimizer for alpha.
+            actor_optimizer (torch.optim.optimizer): The optimizer for actor.
```
Comment (emailweixu) [inline alf/algorithms/sac_algorithm.py]: add back the comment for dqdq_clipping

File: alf/algorithms/sac_algorithm.py [hunk 3]
```diff
@@ -119,23 +126,18 @@ def __init__(self,
...
-        critic_network2 = critic_network.copy(name='CriticNetwork2')
-        log_alpha = tfa_common.create_variable(
-            name='log_alpha',
-            initial_value=initial_log_alpha,
-            dtype=tf.float32,
-            trainable=True)
+        critic_network1 = critic_network.copy()
+        critic_network2 = critic_network.copy()
+
+        log_alpha = nn.Parameter(torch.Tensor([initial_log_alpha]))
```
Comment (emailweixu) [inline alf/algorithms/sac_algorithm.py]: `float(initial_log_alpha)` in case that an int value is provided

File: alf/utils/common.py [hunk 4]
```diff
@@ -129,6 +131,38 @@ def as_list(x):
     return [x]
 
 
+class Periodically(nn.Module):
+    def __init__(self, body, period, name='periodically'):
+        """Periodically performs the ops defined in body.
+
+        Args:
+        body: callable to be performed every time
```
Comment (emailweixu) [inline alf/utils/common.py]: comment style. Indent and arg (type)

File: alf/utils/common.py [hunk 5]
```diff
@@ -129,6 +131,38 @@ def as_list(x):
...
+            an internal counter is divisible by the period.
+        period (int): inverse frequency with which to perform the operation.
+        name: name of the object.
+
+        Raises:
+        TypeError: if body is not a callable.
+
+        Returns:
+        An op that periodically performs the specified body operation.
```
Comment (emailweixu) [inline alf/utils/common.py]: no Returns

File: alf/utils/common.py [hunk 6]
```diff
@@ -854,18 +881,13 @@ def decorate(f):
 def set_random_seed(seed):
     """Set a seed for deterministic behaviors.
 
-    Note: If someone runs an experiment with a pre-selected manual seed, he can
-    definitely reproduce the results with the same seed; however, if he runs the
-    experiment with seed=None and re-run the experiments using the seed previously
-    returned from this function (e.g. the returned seed might be logged to
-    Tensorboard), and if cudnn is used in the code, then there is no guarantee
-    that the results will be reproduced with the recovered seed.
-
```
Comment (emailweixu) [inline alf/utils/common.py]: should not remove

File: alf/utils/common.py [hunk 7]
```diff
@@ -854,18 +881,13 @@ def decorate(f):
...
     Args:
         seed (int|None): seed to be used. If None, a default seed based on
             pid and time will be used.
     Returns:
-        The seed being used if `seed` is None.
+        The seed being used if `seed` is None. Note that you may not reproduce
```
Comment (emailweixu) [inline alf/utils/common.py]: should not change

File: alf/utils/common.py [hunk 8]
```diff
@@ -886,3 +908,53 @@ def log_metrics(metrics, prefix=''):
     """
     log = ['{0} = {1}'.format(m.name, m.result()) for m in metrics]
     logging.info('%s \n\t\t %s', prefix, '\n\t\t '.join(log))
+
+
+def assert_members_are_not_overridden(base_cls,
```
Comment (emailweixu) [inline alf/utils/common.py]: not used. remove

File: alf/algorithms/sac_algorithm_test.py (hunk 1 above)
Comment (Haichao-Zhang) [reply alf/algorithms/sac_algorithm_test.py]: updated

File: alf/algorithms/sac_algorithm.py (hunk 2 above)
Comment (Haichao-Zhang) [reply alf/algorithms/sac_algorithm.py]: Added

File: alf/algorithms/sac_algorithm.py (hunk 3 above)
Comment (Haichao-Zhang) [reply alf/algorithms/sac_algorithm.py]: Added

File: alf/utils/common.py (hunk 4 above)
Comment (Haichao-Zhang) [reply alf/utils/common.py]: Updated

File: alf/utils/common.py (hunk 5 above)
Comment (Haichao-Zhang) [reply alf/utils/common.py]: Changed

File: alf/utils/common.py (hunk 6 above)
Comment (Haichao-Zhang) [reply alf/utils/common.py]: Seems to be caused by a merge. Changed back.

File: alf/utils/common.py (hunk 7 above)
Comment (Haichao-Zhang) [reply alf/utils/common.py]: Changed back.

File: alf/utils/common.py (hunk 6 above)
Comment (Haichao-Zhang) [reply alf/utils/common.py]: There are a number of unintended changes in ```common.py``` due to a merge with PR_batch_env_temp for testing. They are changed back.

File: alf/utils/common.py (hunk 8 above)
Comment (Haichao-Zhang) [reply alf/utils/common.py]: This is also removed.

File: alf/algorithms/sac_algorithm.py [hunk 9]
```diff
@@ -361,9 +361,9 @@ def calc_loss(self, training_info: TrainingInfo):
         return LossInfo(
             loss=actor_loss.loss + critic_loss.loss + alpha_loss.loss,
             extra=SacLossInfo(
-                actor=actor_loss.extra.view(-1).mean(),
-                critic=critic_loss.extra.view(-1).mean(),
-                alpha=alpha_loss.extra.view(-1).mean()))
+                actor=actor_loss.extra,
+                critic=critic_loss.extra,
```
Comment (Haichao-Zhang) [inline alf/algorithms/sac_algorithm.py]: The previous mis-match on critic loss is because of this ```mean()```

**Comment by Haichao-Zhang** [comment]
Now all curves matches, even though the initialization is different, suggesting it is not crucial for the performance.

File: alf/algorithms/sac_algorithm.py [hunk 10]
```diff
@@ -224,63 +219,54 @@ def _actor_train_step(self, exp: Experience, state: SacActorState,
...
-                network_state=state.critic2)
+                exp.observation, state=state.critic2)
 
-            assert isinstance(action_distribution,
-                              tfp.distributions.Categorical), (
-                                  "Only `tfp.distributions.Categorical` " +
-                                  "was supported, received:" + str(
-                                      type(action_distribution)))
+            assert isinstance(
+                action_distribution, td.categorical.Categorical) or \
```
Comment (emailweixu) [inline alf/algorithms/sac_algorithm.py]: use this instead: https://github.com/HorizonRobotics/alf/blob/4723afdba6329459cc16f3204ae901117c86aa05/alf/utils/summary_utils.py#L283

File: alf/algorithms/sac_algorithm.py [hunk 11]
```diff
@@ -224,63 +219,54 @@ def _actor_train_step(self, exp: Experience, state: SacActorState,
...
-            target_q_value = tf.stop_gradient(tf.minimum(critic1, critic2))
-            alpha = tf.stop_gradient(tf.exp(self._log_alpha))
-            actor_loss = tf.reduce_mean(
-                action_probs * (alpha * log_action_probs - target_q_value),
-                axis=-1)
+            target_q_value = torch.min(critic1, critic2).detach()
+            alpha = torch.exp(self._log_alpha)
+            actor_loss = torch.exp(log_action_probs) * (
+                alpha.detach() * log_action_probs - target_q_value)
+            actor_loss = actor_loss.view(actor_loss.size(0), -1).mean(1)
```
Comment (emailweixu) [inline alf/algorithms/sac_algorithm.py]: `view()` may not always work. Either avoid it or use reshape:
```python
actor_loss.mean(list(range(1, actor_loss.ndim)))
```

File: alf/algorithms/sac_algorithm.py [hunk 12]
```diff
@@ -224,63 +219,54 @@ def _actor_train_step(self, exp: Experience, state: SacActorState,
...
+            critic1, critic1_state = self._critic_network1(
+                critic_input, state=state.critic1)
+            critic2, critic2_state = self._critic_network2(
+                critic_input, state=state.critic2)
 
-                target_q_value = tf.minimum(critic1, critic2)
+            target_q_value = torch.min(critic1, critic2)
 
-            dqda = tape.gradient(target_q_value, action)
+            dqda = torch.autograd.grad(target_q_value, action,
```
Comment (emailweixu) [inline alf/algorithms/sac_algorithm.py]: action might be nested.

File: alf/algorithms/sac_algorithm.py (hunk 10 above)
Comment (Haichao-Zhang) [reply alf/algorithms/sac_algorithm.py]: Moved it to ```dist_utils.py``` as it is generally useful.

File: alf/algorithms/sac_algorithm.py (hunk 11 above)
Comment (Haichao-Zhang) [reply alf/algorithms/sac_algorithm.py]: Changed according to suggestion.

File: alf/algorithms/sac_algorithm.py (hunk 12 above)
Comment (Haichao-Zhang) [reply alf/algorithms/sac_algorithm.py]: Good point. Changed to properly handle nested action.

File: alf/algorithms/sac_algorithm.py [hunk 13]
```diff
@@ -226,10 +226,10 @@ def _actor_train_step(self, exp: Experience, state: SacActorState,
 
             target_q_value = torch.min(critic1, critic2)
 
-            dqda = torch.autograd.grad(target_q_value, action,
-                                       torch.ones(target_q_value.size()))[0]
-
-            def actor_loss_fn(dqda, action):
+            def actor_loss_fn(action):
```
Comment (emailweixu) [inline alf/algorithms/sac_algorithm.py]: It's better to do grad only once if possible. The following might work:
```python
        dqda = alf.nest.pack_sequence_as(action,
            torch.autograd.grad(target_q_value, alf.nest.flatten(action)))
```

Comment (Haichao-Zhang) [reply alf/algorithms/sac_algorithm.py]: Now changed to one pass computation as suggested.

**Pull request merged**